src/Chainweb/Rosetta/Utils.hs

```diff
@@ -182,15 +207,644 @@ toContNextStep currChainId pe
+sigToScheme RosettaEd25519 = pure P.ED25519
+sigToScheme st = Left $ stringRosettaError RosettaInvalidSignature $
+                 "Found unsupported SignatureType: " ++ show st
+
+
+newtype AccountName = AccountName { _accountName :: T.Text }
+  deriving (Show, Eq, Ord)
+instance Hashable AccountName where
+  hash (AccountName n) = hash n
+  hashWithSalt i (AccountName n) = hashWithSalt i n
+
+-- TODO: If AccountId metadata changes to include the account guard,
+-- will need to ask for keyset here.
+acctNameToAcctId :: AccountName -> AccountId
+acctNameToAcctId (AccountName name) = accountId name
+
+
+--------------------------------------------------------------------------------
+-- /payloads
+
+data PayloadsMetaData = PayloadsMetaData
+  { _payloadsMetaData_signers :: ![(Signer, AccountId)]
+  , _payloadsMetaData_nonce :: !T.Text
+  , _payloadsMetaData_publicMeta :: !P.PublicMeta
+  , _payloadsMetaData_tx :: !ConstructionTx
+  -- ^ Needed to construct gas payer AccountId
+  } deriving (Show)
+instance ToObject PayloadsMetaData where
+  toPairs (PayloadsMetaData signers nonce pm tx) =
+    [ "signers" .= signers
+    , "nonce" .= nonce
+    , "public_meta" .= pm
+    , "tx" .= tx
+    ]
+  toObject m = HM.fromList (toPairs m)
+instance FromJSON PayloadsMetaData where
+  parseJSON = withObject "PayloadsMetaData" $ \o -> do
+    signers <- o .: "signers"
+    nonce <- o .: "nonce"
+    publicMeta <- o .: "public_meta"
+    tx <- o .: "tx"
+    pure PayloadsMetaData
+      {  _payloadsMetaData_signers = signers
+      , _payloadsMetaData_nonce = nonce
+      , _payloadsMetaData_publicMeta = publicMeta
+      , _payloadsMetaData_tx = tx
+      }
+
+
+data EnrichedCommand = EnrichedCommand
+  { _enrichedCommand_cmd :: !(Command T.Text)
+  , _enrichedCommand_txInfo :: !ConstructionTx
+  , _enrichedCommand_signerAccounts :: ![AccountId]
+  } deriving (Show)
+instance ToJSON EnrichedCommand where
+  toJSON (EnrichedCommand cmd tx accts) = object
+    [ "cmd" .= cmd
+    , "tx_info" .= tx
+    , "signer_accounts" .= accts ]
+instance FromJSON EnrichedCommand where
+  parseJSON = withObject "EnrichedCommand" $ \o -> do
+    cmd <- o .: "cmd"
+    txInfo <- o .: "tx_info"
+    accts <- o .: "signer_accounts"
+    pure EnrichedCommand
+      { _enrichedCommand_cmd = cmd
+      , _enrichedCommand_txInfo = txInfo
+      , _enrichedCommand_signerAccounts = accts
+      }
+
+
+enrichedCommandToText :: EnrichedCommand -> T.Text
+enrichedCommandToText = T.decodeUtf8 . BSL.toStrict . encode
+
+textToEnrichedCommand :: T.Text -> Maybe EnrichedCommand
+textToEnrichedCommand = decodeStrict' . T.encodeUtf8
+
+
+constructionTxToPactRPC
+    :: ConstructionTx
+    -> P.PactRPC T.Text
+constructionTxToPactRPC txInfo =
+  case txInfo of
+    ConstructTransfer from _ to toGuard amt ->
+      let code = t $! printf
+            "(coin.transfer-create \"%s\" \"%s\" (read-keyset \"%s\") (read-decimal \"%s\"))"
```

Comment:
Emily being paranoid: potential tx injection here. Let's triple check.

Reply:
Good catch! I proposed a solution in the latest commit (it uses `show` to get the escape string)